### ISSUE TITLE
Get-Vegetable gives no output

### ISSUE THREAD

**Issue by DCFinley** [jdhitsolutions/PSTeachingTools#4]
I downloaded and installed PSTeachingTools module from Repository, but running Get-Vegetable produced no result. I re-ran with -verbose and saw that the Process produced no results and was terminated with End.

I uninstalled PSTeachingTools, then re-downloaded/installed from Repository with -verbose. Towards the end I saw this:
"Catalog file 'PSTeachingTools.cat' is not found in the contents of the module 'PSTeachingTools' being installed."

I think this may be why I received no results for Get-Vegetable after the install. Can you please verify?

**Comment by jdhitsolutions** [comment]
That's very odd. I just installed on a new system with no errors. This sounds like an issue with PowerShellGet. What version are you running?

```powershell
Get-Module PowerShellGet
```
I'm assuming when you run Find-Module you get this result:

```powershell
[SRV1]: PS C:\Users\artd\Documents> find-module psteachingtools

Version    Name                                Repository           Description
-------    ----                                ----------           -----------
3.2.0      PSTeachingTools                     PSGallery            A set of commands and tools for teaching PowerSh...
```
And just in case,what operating system and PowerShell version are you running?

**Comment by jdhitsolutions** [comment]
I uninstalled the module and re-installed using -Verbose. I see the message you do.

```powershell
 install-module psteachingtools -verb
VERBOSE: Using the provider 'PowerShellGet' for searching packages.
VERBOSE: The -Repository parameter was not specified.  PowerShellGet will use all of the registered repositories.       VERBOSE: Getting the provider object for the PackageManagement Provider 'NuGet'.                                        VERBOSE: The specified Location is 'https://www.powershellgallery.com/api/v2' and PackageManagementProvider is 'NuGet'. VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='psteachingtools'' for ''.VERBOSE: Total package yield:'1' for the specified package 'psteachingtools'.                                           VERBOSE: Performing the operation "Install-Module" on target "Version '3.2.0' of module 'PSTeachingTools'".             
Untrusted repository
You are installing the modules from an untrusted repository. If you trust this repository, change its
InstallationPolicy value by running the Set-PSRepository cmdlet. Are you sure you want to install the modules from
'PSGallery'?
[Y] Yes  [A] Yes to All  [N] No  [L] No to All  [?] Help (default is "N"): y
VERBOSE: The installation scope is specified to be 'AllUsers'.
VERBOSE: The specified module will be installed in 'C:\Program Files\WindowsPowerShell\Modules'.
VERBOSE: The specified Location is 'NuGet' and PackageManagementProvider is 'NuGet'.
VERBOSE: Downloading module 'PSTeachingTools' with version '3.2.0' from the repository 'https://www.powershellgallery.com/api/v2'.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='PSTeachingTools'' for ''.
VERBOSE: InstallPackage' - name='PSTeachingTools', version='3.2.0',destination='C:\Users\artd\AppData\Local\Temp\295789385'
VERBOSE: DownloadPackage' - name='PSTeachingTools', version='3.2.0',destination='C:\Users\artd\AppData\Local\Temp\295789385\PSTeachingTools\PSTeachingTools.nupkg', uri='https://www.powershellgallery.com/api/v2/package/PSTeachingTools/3.2.0'
VERBOSE: Downloading 'https://www.powershellgallery.com/api/v2/package/PSTeachingTools/3.2.0'.
VERBOSE: Completed downloading 'https://www.powershellgallery.com/api/v2/package/PSTeachingTools/3.2.0'.
VERBOSE: Completed downloading 'PSTeachingTools'.
VERBOSE: Hash for package 'PSTeachingTools' does not match hash provided from the server.
VERBOSE: InstallPackageLocal' - name='PSTeachingTools', version='3.2.0',destination='C:\Users\artd\AppData\Local\Temp\295789385'
VERBOSE: Catalog file 'PSTeachingTools.cat' is not found in the contents of the module 'PSTeachingTools' being installed.
VERBOSE: Module 'PSTeachingTools' was installed successfully to path 'C:\Program Files\WindowsPowerShell\Modules\PSTeachingTools\3.2.0'.
```

What is your execution policy? Is your computer managed by GroupPolicy or AppLocker?

**Comment by DCFinley** [comment]
Good morning, Jeff!

I'm using Version 1.0.0.1 of PowerShellGet.

RE: execution policy, Local Machine is Remote Signed, and all other Scopes
are Undefined.

I'm unsure re: what manages my computer. I've only had it a few weeks and
got it to take classes. Yes I'm a noob. :)

Thanks.

**Comment by jdhitsolutions** [comment]
Can you install and run other modules from the PowerShell Gallery? I get the same verbose messages regardless of what I'm installing. The lack of the cat file is just informative. The module doesn't have one. See if you can download the PSScriptTools module and import it into your PowerShell session. Try to run a command.

```powershell
Install-Module PSScriptTools -force
Import-Module PSScriptTools
Get-WindowsVersion
```

Then we'll have a better idea if the problem is more than just the PSTeachingTools module.

**Comment by jdhitsolutions** [comment]
Assuming other modules work, or maybe even if they don't, run these commands in your PowerShell session.

```powershell
start-transcript $home\documents\log.txt
get-module psteachingtools -list
get-command -module psteachingtools
get-vegetable -verbose
stop-transcript
```

This will create a transcript file called log.txt in your Documents folder. You can redact the user name but everything else is just version numbers. Paste the file contents here. This will let me see exactly what you are typing and the results you are getting.

**Comment by DCFinley** [comment]
Transcript run, commands run, text results below:
**********************
Windows PowerShell transcript start
Start time: 20200506060421
Username: DESKTOP-U2SLRU4\dclar
RunAs User: DESKTOP-U2SLRU4\dclar
Configuration Name:
Machine: DESKTOP-U2SLRU4 (Microsoft Windows NT 10.0.18363.0)
Host Application: C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe
Process ID: 8652
PSVersion: 5.1.18362.752
PSEdition: Desktop
PSCompatibleVersions: 1.0, 2.0, 3.0, 4.0, 5.0, 5.1.18362.752
BuildVersion: 10.0.18362.752
CLRVersion: 4.0.30319.42000
WSManStackVersion: 3.0
PSRemotingProtocolVersion: 2.3
SerializationVersion: 1.1.0.1
**********************
Transcript started, output file is C:\Users\dclar\documents\log.txt
PS C:\windows\system32> get-module psteachingtools -list


    Directory: C:\Program Files\WindowsPowerShell\Modules


ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     3.2.0      PSTeachingTools                     {Get-Vegetable,
New-Vegetable, Set-Vegetable, Start-TypedD...


PS C:\windows\system32> get-command -module psteachingtools

CommandType     Name                                               Version
   Source
-----------     ----                                               -------
   ------
Function        Get-Vegetable                                      3.2.0
   PSTeachingTools
Function        New-Vegetable                                      3.2.0
   PSTeachingTools
Function        Set-Vegetable                                      3.2.0
   PSTeachingTools
Function        Start-TypedDemo                                    3.2.0
   PSTeachingTools


PS C:\windows\system32> get-vegetable -verbose
VERBOSE: [BEGIN  ] Starting: Get-Vegetable
VERBOSE: [PROCESS] Getting all vegetables
VERBOSE: [END    ] Ending: Get-Vegetable
PS C:\windows\system32> stop-transcript
**********************
Windows PowerShell transcript end
End time: 20200506060552
**********************

Also, I attached a screenshot of the error message I received when I tried
running Get-WindowVersion.

**Comment by jdhitsolutions** [comment]
The command is `Get-WindowsVersion`. Your screen shot didn't post.

**Comment by DCFinley** [comment]
Sorry, typo in my last email, but I did run Get-WindowsVersion.

Re-attached screen shot and also inserted into email.

[image: Screenshot (2).png]

**Comment by jdhitsolutions** [comment]
Guess I should look at email today.

**Comment by jdhitsolutions** [comment]
Saw your email, but attachment must have been stripped off somewhere.  `Get-WindowsVersion` should have given you something like this:

```

   Computername: BOVINE320

ProductName                    EditionID            ReleaseID  Build  InstalledUTC
-----------                    ---------            ---------  -----  ------------
Windows 10 Pro                 Professional         1909       18363  7/5/2019 10:54:49 PM
```

Anything else, just copy and paste the result.

**Comment by jdhitsolutions** [comment]
Can you run this after importing the PSTeachingTools module?

```powershell
PS C:\> [vegetable]::new()

UPC     Count Name                 State    Color
---     ----- ----                 -----    -----
0          17                      Raw      green
```

**Comment by DCFinley** [comment]
Try the new cross-platform PowerShell https://aka.ms/pscore6

PS C:\Users\dclar> get-windowsversion
[DESKTOP-U2SLRU4] Connecting to remote server DESKTOP-U2SLRU4 failed with
the following error message : WinRM cannot
process the request. The following error with errorcode 0x8009030e occurred
while using Negotiate authentication: A
specified logon session does not exist. It may already have been terminated.
 Possible causes are:
  -The user name or password specified are invalid.
  -Kerberos is used when no authentication method and no user name are
specified.
  -Kerberos accepts domain user names, but not local user names.
  -The Service Principal Name (SPN) for the remote computer name and port
does not exist.
  -The client and remote computers are in different domains and there is no
trust between the two domains.
 After checking for the above issues, try the following:
  -Check the Event Viewer for events related to authentication.
  -Change the authentication method; add the destination computer to the
WinRM TrustedHosts configuration setting or
use HTTPS transport.
 Note that computers in the TrustedHosts list might not be authenticated.
   -For more information about WinRM configuration, run the following
command: winrm help config. For more
information, see the about_Remote_Troubleshooting Help topic.
    + CategoryInfo          : OpenError: (DESKTOP-U2SLRU4:String) [],
PSRemotingTransportException
    + FullyQualifiedErrorId : 1312,PSSessionStateBroken
PS C:\Users\dclar>

**Comment by DCFinley** [comment]
Result:
PS C:\Windows\system32> cd c:\
PS C:\> [vegetable]::[new]
Unable to find type [new].
At line:1 char:14
+ [vegetable]::[new]
+              ~~~~~
    + CategoryInfo          : InvalidOperation: (new:TypeName) [],
RuntimeException
    + FullyQualifiedErrorId : TypeNotFound

PS C:\>

**Comment by jdhitsolutions** [comment]
I guess that might have been a bad command to test with because it looks like you don't have remoting enabled.  Does `Get-PSWho` run without error?

**Comment by jdhitsolutions** [comment]
Just realized you tested with the wrong syntax.

```powershell
[vegetable]::new()
```

PowerShell punctuation can get a bit nuts sometimes.

**Comment by DCFinley** [comment]
It does:

User            : DESKTOP-U2SLRU4\dclar
Elevated        : True
Computername    : DESKTOP-U2SLRU4
OperatingSystem : Microsoft Windows 10 Pro [64-bit]
OSVersion       : 10.0.18363
PSVersion       : 5.1.18362.752
Edition         : Desktop
PSHost          : ConsoleHost
WSMan           : 3.0
ExecutionPolicy : RemoteSigned
Culture         : English (United States)



PS C:\>

**Comment by jdhitsolutions** [comment]
Ok. Then we'll assume there is something in PSTeachingtools that doesn't like you. After importing the module does `$VegetablePLU` show anything?

**Comment by DCFinley** [comment]
OMG it shows everything, had to Control-C to stop.

**Comment by jdhitsolutions** [comment]
Let's also make sure you have the right files. You should have something like this:

```powershell
PS C:\> dir ( split-path (get-module psteachingtools).path)


    Directory: C:\Program Files\WindowsPowerShell\Modules\psteachingtools\3.2.0


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----         5/6/2020   4:10 AM                assets
d-----         5/6/2020   4:10 AM                docs
d-----         5/6/2020   4:10 AM                en-US
-a----        4/13/2020   2:41 PM           2581 Changelog.md
-a----        4/13/2020   2:41 PM           1132 LICENSE.txt
-a----        4/13/2020   2:41 PM          45963 plu.csv
-a----        4/13/2020   2:41 PM           7738 PSTeachingTools.psd1
-a----        4/13/2020   2:41 PM            258 PSTeachingTools.psm1
-a----        4/13/2020   2:41 PM           3319 rawveggies.json
-a----        4/13/2020   2:41 PM           1922 README.md
-a----        4/13/2020   2:41 PM          12473 Start-TypedDemo.ps1
-a----        4/13/2020   2:41 PM           1007 vegetable-class.ps1
-a----        4/13/2020   2:41 PM           3676 vegetable.format.ps1xml
-a----        4/13/2020   2:41 PM           5292 Vegetables.ps1
```

**Comment by jdhitsolutions** [comment]
Then you should also have a `$myvegetables` object after importing the module. This is the source for `Get-Vegetable`.

And you are on a normal desktop right? You aren't logging into some VDI or cloud-based system are you? I've had plenty of people install this module so I'm very puzzled about what is different with o you that is causing a problem.

**Comment by DCFinley** [comment]
This is odd. I just retried the Get-Vegetable command - and it worked:

4372    Retailer Assigned (includes pre-cut melons) Melon
4373    Retailer Assigned (includes pre-cut melons) Melon
4374    Retailer Assigned (includes pre-cut melons) Melon
4375    Retailer Assigned (includes pre-cut melons) Melon
4376    Retailer Assigned (includes pre-cut melons) Melon
4377    Small Yellow Flesh, Tree Ripened, Ready-to-eat Nectarine
4378    Large Yellow Flesh, Tree Ripened, Ready-to-eat Nectarine
*PS C:\> get-vegetable*

UPC     Count Name                 State    Color
---     ----- ----                 -----    -----
4078       12 corn                 Roasted  yellow
4064        4 tomato               Raw      red
4062       11 cucumber             Raw      green
4562       10 carrot               Raw      orange
4089       13 radish               Raw      red
4674       14 peas                 Steamed  green
4811       12 turnip               Boiled   purple
4725       18 russet potato        Fried    brown
4060       15 broccoli             Steamed  green
4067        7 zucchini             Raw      green
4090        4 spinach              Raw      green
4572        7 cauliflower          Steamed  white
3125       17 habanero pepper      Raw      orange
4677       16 Anaheim pepper       Raw      green
4088       19 red bell pepper      Sauteed  red
4081        6 eggplant             Fried    purple
4604        2 endive               Raw      green

**Comment by jdhitsolutions** [comment]
Hmmm. Open a brand new PowerShell session and try again.

**Comment by DCFinley** [comment]
From the new session:
PS C:\Windows\system32> get-vegetable

UPC     Count Name                 State    Color
---     ----- ----                 -----    -----
4078       12 corn                 Roasted  yellow
4064        4 tomato               Raw      red
4062       11 cucumber             Raw      green
4562       10 carrot               Raw      orange
4089       13 radish               Raw      red
4674       14 peas                 Steamed  green
4811       12 turnip               Boiled   purple
4725       18 russet potato        Fried    brown
4060       15 broccoli             Steamed  green
4067        7 zucchini             Raw      green
4090        4 spinach              Raw      green
4572        7 cauliflower          Steamed  white
3125       17 habanero pepper      Raw      orange
4677       16 Anaheim pepper       Raw      green
4088       19 red bell pepper      Sauteed  red
4081        6 eggplant             Fried    purple
4604        2 endive               Raw      green


PS C:\Windows\system32>

I'm stumped. But then, I'm new, so that's my default.

**Comment by jdhitsolutions** [comment]
Being new doesn't come into play here at all. You were running the commands as I expected and you should have seen output from the very beginning. That's why I kept thinking it was something about your Windows installation or even hardware. Maybe antivirus was getting in the way. 

When you looked at `$vegetableplu` even though there are 1400 items, it should only have taken a second to display.  If it was taking such a long time that you felt the need to Ctrl+C out of it, something is going on. Run `$VegetablePLU` again and let it finish. Then see how long it took to run. It should be like this:

```powershell
PS C:\> get-history -count 1 | select *


Id                 : 32
CommandLine        : $vegetableplu
ExecutionStatus    : Completed
StartExecutionTime : 5/6/2020 12:54:44 PM
EndExecutionTime   : 5/6/2020 12:54:45 PM
```

**Comment by DCFinley** [comment]
Results:

PS C:\Windows\system32> get-history -count 1 | select *


Id                 : 2
CommandLine        : $vegetablePLU
ExecutionStatus    : Completed
StartExecutionTime : 5/6/2020 10:04:38 AM
EndExecutionTime   : 5/6/2020 10:07:34 AM

**Comment by jdhitsolutions** [comment]
That is much slower than I would have expected. 3 minutes sounds like something is being downloaded and then displayed. That would explain why Get-Vegetable failed to give you any results. Do you know if you are running anything from the cloud or in a virtual machine or some other type of sandbox? Could your computer be on the low side of memory and processor?

In Windows PowerShell if you run `Get-WmiObject win32_computersystem` how much totalphysicalmemory do you see?

**Comment by jdhitsolutions** [comment]
Or here's another test. Run `1..100` which should display 1 through 100. This should take about a second. If it takes longer something is most definitely up with your system. At least compared to what I think should be considered normal.

**Comment by DCFinley** [comment]
It's 16GB total, but using 15.5:

PS C:\Windows\system32> get-wmiobject win32_computersystem


Domain              : WORKGROUP
Manufacturer        : LENOVO
Model               : 81H5
Name                : DESKTOP-U2SLRU4
PrimaryOwnerName    : dclarkfinley@gmail.com
TotalPhysicalMemory : 16608780288

I ran 1.100, then did this get-history -count:

PS C:\Windows\system32> get-history -count 1 | select *


Id                 : 9
CommandLine        : 1..100
ExecutionStatus    : Completed
StartExecutionTime : 5/6/2020 10:34:11 AM
EndExecutionTime   : 5/6/2020 10:34:23 AM

12 seconds,

**Comment by jdhitsolutions** [comment]
That would explain a lot. If your computer is using that much memory, I'm betting CPU usage is also high and that's why PowerShell is so slow to run which would explain the weird errors.

**Comment by DCFinley** [comment]
I misspoke, inasmuch as total available memory is 15.5GB. That said,
according to Task Manager 12TB is being used.

**Comment by jdhitsolutions** [comment]
Try these commands which will give us a better idea.

```powershell
get-wmiobject win32_operatingsystem | Select Caption,*memory*,OSArch*
get-wmiobject win32_processor | Select Caption,LoadPercentage
```

Based on the Lenovo model number you appear to be running an older IdeaPad.

**Comment by DCFinley** [comment]
Hi Jeff,

Sorry for the delay, life intervened...

Results:
PS C:\Windows\system32> get-wmiobject win32_operatingsystem | Select
Caption, *memory*, osarch*


Caption                : Microsoft Windows 10 Pro
FreePhysicalMemory     : 3623384
FreeVirtualMemory      : 4814248
MaxProcessMemorySize   : 137438953344
TotalVirtualMemorySize : 18709880
TotalVisibleMemorySize : 16219512
OSArchitecture         : 64-bit


PS C:\Windows\system32> get-wmiobject win32_processor | Select
Caption,LoadPercentage

Caption                              LoadPercentage
-------                              --------------
AMD64 Family 21 Model 112 Stepping 0             26

I then:
1) Rebooted my computer;
2) Opened Task Manager, discovered what a memory hog Adobe Creative Suite
is and shut it down;
3) Closed unnecessary Chrome windows.
4) Available memory improved to 4.8GB or 68%, so I opened a PowerShell
Admin console and gave Get-Vegetable command. It worked fine.

Thanks so much for spending so much time with me today on this. It's a bit
embarrasing to realize that it's on me, or at least on the weak computer
the school purchased for me for this class. I'm taking an Azure Admin
course, coming into it from the project management POV but I'm now enjoying
learning the tech far more that I ever did managing. As Azure is so
PowerShell-heavy I'm using your courses to get me up to speed, and they're
great!

Thanks again,
Dave Finley


PS: As I'm also a newb to GitHub, do I close that issue, or do you?

**Comment by jdhitsolutions** [comment]
I can close it. I'm glad we got this working for you. You were clearly pushing that machine to do too much. Good luck.
